File: lib/py/mos_algorithm.py
import math

def mos_algorithm(n, arr, queries):
    sqrt_n = int(math.sqrt(n))
    
    # 정렬 (Hilbert variant or even/odd variant)
    def query_sort_key(q):
        l, r, idx = q
        block = l // sqrt_n
        return (block, r if block % 2 == 1 else -r)
        
    queries.sort(key=query_sort_key)
    
    ans = [0] * len(queries)
    cnt = [0] * 1000005
    res = 0
    
    def add(idx):
        nonlocal res
        cnt[arr[idx]] += 1
        if cnt[arr[idx]] == 1:
            res += 1
            
    def remove(idx):
        nonlocal res
        cnt[arr[idx]] -= 1
        if cnt[arr[idx]] == 0:
            res -= 1
            
    l, r = queries[0][0], queries[0][0] - 1
    
    for q_l, q_r, q_idx in queries:
        while l > q_l:
            l -= 1
            add(l)
        while r < q_r:
            r += 1
            add(r)
        while l < q_l:
            remove(l)
            l += 1
        while r > q_r:
            remove(r)
            r -= 1
        ans[q_idx] = res
        
    return ans

File: lib/cpp/mos_algorithm.cpp
#include <algorithm>
#include <cmath>
#include <vector>


using namespace std;

struct Query {
  int l, r, idx;
};

int sqrtN;

bool compare(const Query &a, const Query &b) {
  if (a.l / sqrtN != b.l / sqrtN)
    return a.l / sqrtN < b.l / sqrtN;
  return (a.l / sqrtN) % 2 == 1
             ? a.r < b.r
             : a.r > b.r; // Hilbert Curve optimization variant
}

// 전역 변수들 (문제에 따라 수정)
int cnt[1000005];
int res = 0;
vector<int> arr;

void add(int idx) {
  if (++cnt[arr[idx]] == 1)
    res++; // 예: 서로 다른 수 개수
}

void remove(int idx) {
  if (--cnt[arr[idx]] == 0)
    res--;
}

vector<int> mos_algorithm(int n, const vector<int> &input_arr,
                          vector<Query> &queries) {
  arr = input_arr;
  sqrtN = sqrt(n);
  sort(queries.begin(), queries.end(), compare);

  vector<int> ans(queries.size());
  int l = queries[0].l, r = queries[0].l - 1;

  for (const auto &q : queries) {
    while (l > q.l)
      add(--l);
    while (r < q.r)
      add(++r);
    while (l < q.l)
      remove(l++);
    while (r > q.r)
      remove(r--);
    ans[q.idx] = res;
  }
  return ans;
}
